### ISSUE TITLE
`useless-else-on-loop` (`PLW0120`) - false positive on possibly empty iterable

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#28382]
### Summary

```py
def fn(values: list[int]):
    for value in values:
        print(value)
    else: # error: `else` clause on loop without a `break` statement; remove the `else` and dedent its contents (useless-else-on-loop)
        print("no values")


fn([])
```
https://play.ruff.rs/7286e06f-334e-489c-8dd6-6b7efb5effd9

[the documentation](https://docs.astral.sh/ruff/rules/useless-else-on-loop/) for this rule says that the `else` will never run if the `for` loop does not contain a `break`, which is not true. the code above outputs `"no values"`. following the suggestion would result in incorrect behavior if the list is not empty.

this rule should only be reported in cases where ruff is able to statically determine that the iterable is not empty (ie. when iterating over a literal value)

### Version

v0.16.6

**Comment by vinitsonawane45** [comment]
@DetachHead I'd like to work on this issue.

My understanding is that `PLW0120` currently reports a `for` loop's `else` clause whenever there is no `break`, but this can be a false positive when the iterable may be empty:

```python
def fn(values: list[int]):
    for value in values:
        print(value)
    else:
        print("no values")
```

Here, the `else` clause is reached when `values` is empty, so removing it can change the program's behavior.

Before I start implementing a fix, I'd like to confirm the intended design with the maintainers. My initial thought is that the diagnostic should only be emitted when Ruff can statically determine that the iterable is guaranteed to be non-empty, rather than simply checking for the absence of `break`.

Would that be the intended direction, or would you prefer a narrower approach, such as only handling statically known non-empty iterables?

**Comment by MichaReiser** [comment]
I'm probably missing something but

> [the documentation](https://docs.astral.sh/ruff/rules/useless-else-on-loop/) for this rule says that the else will never run if the for loop does not contain a break,

My reading of the documentation is that it says the opposite:

> If a loop always terminates "normally" (i.e., does not contain a break), then the else clause is redundant, as the code inside the else clause will *always* be executed.


Which you can see using your reproducer:

```pycon
>>> fn([])
no values
>>> fn([1])
1
no values
```

And the rule suggests you change the code to

```py
def fn(values: list[int]):
    for value in values:
        print(value)
    print("no values")
```

**Comment by DetachHead** [comment]
uhh oops, i keep misremembering how `for`/`else` works it's so unintuitive... at this point i would prefer a rule to just ban the syntax altogether lol

**Comment by MichaReiser** [comment]
>  i keep misremembering how for/else works it's so unintuitive...

I agree on this. It's also the reason why I felt very unsure whether I'm missing something, especially because you authored the issue and KotlinIsland upvoted it.
